src/Microsoft.DotNet.Archive/IndexedArchive.cs

```diff
@@ -0,0 +1,583 @@
+                            }
+                            else
+                            {
+                                string destinationPath = Path.Combine(outputDirectory, target);
+                                extractOperations.Add(new CopyOperation(extractSource, destinationPath));
+                            }
+                        }
+                    }
+
+                    int opsExecuted = 0;
+                    // execute all operations
+                    //foreach(var extractOperation in extractOperations)
+                    extractOperations.AsParallel().ForAll(extractOperation =>
+                    {
+                        extractOperation.DoOperation();
+                        progress.Report("Expanding", Interlocked.Increment(ref opsExecuted), extractOperations.Count);
+                    });
+                }
+            }
+        }
+
+        public void AddExternalDirectory(string externalDirectory)
+        {
+            CheckDisposed();
+            foreach (var externalFile in Directory.EnumerateFiles(externalDirectory, "*", SearchOption.AllDirectories))
+            {
+                AddExternalFile(externalFile);
+            }
+        }
+
+        public void AddExternalFile(string externalFile)
+        {
+            CheckDisposed();
+            using (var fs = File.OpenRead(externalFile))
+            {
+                string hash = GetHash(fs); 
+                // $ prefix indicates that the file is not in the archive and path is relative to an external directory
+                archiveFiles[hash] = new ArchiveFileInfo(null, "$" + hash , hash);
+                externalFiles[hash] = externalFile;
+            }
+        }
+        public void AddDirectory(string sourceDirectory, IProgress<ProgressReport> progress, string destinationDirectory = null)
+        {
+            var sourceFiles = Directory.EnumerateFiles(sourceDirectory, "*", SearchOption.AllDirectories).ToArray();
+            int filesAdded = 0;
+            sourceFiles.AsParallel().ForAll(sourceFile =>
+                {
+                    string destinationPath = sourceFile.Substring(sourceDirectory.Length + 1);
+
+                    if (destinationDirectory != null)
+                    {
+                        destinationPath = Path.Combine(destinationDirectory, destinationPath);
+                    }
+
+                    string extension = Path.GetExtension(sourceFile);
+
+                    if (ZipExtensions.Any(ze => ze.Equals(extension, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        AddZip(sourceFile, destinationPath);
+                    }
+                    else
+                    {
+                        AddFile(sourceFile, destinationPath);
+                    }
+
+                    progress.Report($"Adding {sourceDirectory}", ++filesAdded, sourceFiles.Length);
```

Comment:
I'm not too familiar with Parallel LINQ but is `filesAdded` here going to get locked so it's rcorrectly?